File: src/naoqi_libqicore/libqicore/example/file/file_example.py

import sys
import qi
import qi.path
import time


app = qi.Application()
qicore = qi.module('qicore')

log = qi.Logger("qicore.file.example")

def print_tranfer_progress(progress_value):
  log.info(">>>> File Transfer Progress = {0}%".format(progress_value * 100))

def work_on_image_file(image_file_path):
  log.info("Working on image file at {0} ...".format(image_file_path))
  # we fake working on it...
  time.sleep(1)
  log.info("Working on image file at {0} - DONE".format(image_file_path))


def store_image(image_store, name, image_file_path):

  log.info("Storing image {1} at {0} into the ImageStore...".format(image_file_path, name))

  # First open the file with read-only shareable access.
  file = qicore.openLocalFile(image_file_path)

  # Now we can share reading access to this file.
  image_store.store_image(file, name)

  log.info("Storing image {1} at {0} into the ImageStore - DONE".format(image_file_path, name))


def process_image(image_store, image_file_name, image_file_path):

  # We acquire read-only access to the file and retrieve it locally.
  file = image_store.get_image(image_file_name)
  qicore.copyToLocal(file, image_file_path)

  # We don't need the remote access anymore.
  del file

  # Now work on the file located at `fileLocation`.
  work_on_image_file(image_file_path)


def process_image_with_progress(image_store, image_file_name, image_file_path):

  # We acquire read-only access to the file.
  file = image_store.get_image(image_file_name)

  # We prepare the operation without launching it yet:
  file_operation = qicore.FileCopyToLocal(file, image_file_path)

  # We want to see the progress so we plug a logging function.
  file_operation.notifier().progress.connect(print_tranfer_progress)

  # Launch the copy and wait for it to end before continuing.
  file_operation.start() # In real projects, prefer file_operation.start(_async=True).then(nextFunction) instead.

  # We don't need the remote access anymore.
  del file
  del file_operation

  # Now work on the file located at `fileLocation`.
  work_on_image_file(image_file_path)


def do_some_work(client_session, image_path, image_name):

  image_store = client_session.service('ImageStore')
  assert image_store
  store_image(image_store, image_name, image_path)
  process_image(image_store, image_name, "./tempfile")
  process_image_with_progress(image_store, image_name, "./tempfile")



path_to_file = sys.argv[1]
assert path_to_file

file_id = sys.argv[2]
assert file_id

app.start()

do_some_work(app.session, path_to_file, file_id)

File: src/naoqi_libqicore/libqicore/example/file/file_example.cpp
#include <qicore/file.hpp>
#include <qi/session.hpp>
#include <qi/application.hpp>
#include <qi/type/dynamicobjectbuilder.hpp>
#include <testsession/testsessionpair.hpp>
#include <testsession/testsession.hpp>

#include "imagestore.hpp"

qiLogCategory("test.qi.fileExample");

namespace bob
{
void printTranferProgress(double progress)
{
  qiLogInfo() << ">>>> File Transfer Progress = " << (progress * 100.0) << "%";
}

void workOnImageFile(const qi::Path& imageFilePath)
{
  qiLogInfo() << "Working on image file at " << imageFilePath << " ...";
  // we fake working on it...
  boost::this_thread::sleep_for(boost::chrono::milliseconds(300));
  qiLogInfo() << "Working on image file at " << imageFilePath << " - DONE";
}

void storeImage(alice::ImageStorePtr imageStore, const std::string& name, const qi::Path& imageFilePath)
{
  qiLogInfo() << "Storing image file at " << imageFilePath << " into the ImageStore...";

  // First open the file with read-only shareable access.
  qi::FilePtr file = qi::openLocalFile(imageFilePath);

  // Now we can share reading access to this file.
  imageStore->storeImage(file, name);

  qiLogInfo() << "Storing image file at " << imageFilePath << " into the ImageStore - DONE";
}

void processImage(alice::ImageStorePtr imageStore, const std::string& imageFile, const qi::Path& imageFilePath)
{
  // We acquire read-only access to the file and retrieve it locally.
  qi::FilePtr file = imageStore->getImage(imageFile);
  qi::copyToLocal(file, imageFilePath);

  // We don't need the remote access anymore.
  file.reset();

  // Now work on the file located at `fileLocation`.
  workOnImageFile(imageFilePath);
}

void processImageWithProgress(alice::ImageStorePtr imageStore,
                              const std::string& imageFile,
                              const std::string& imageFilePath)
{
  // We acquire read-only access to the file.
  qi::FilePtr file = imageStore->getImage(imageFile);

  // We prepare the operation without launching it yet:
  qi::FileCopyToLocal fileOp{file, imageFilePath};

  // We want to see the progress so we plug a logging function.
  fileOp.notifier()->progress.connect(&printTranferProgress);

  // Launch the copy and wait for it to end before continuing.
  fileOp.start().wait(); // Don't wait for futures in real code, you should .connect() instead.

  // We don't need the remote access anymore.
  file.reset();

  // Now work on the file located at `fileLocation`.
  workOnImageFile(imageFilePath);
}

void doSomeWork(qi::SessionPtr clientSession, const qi::Path& imagePath, const std::string& imageName)
{
  qi::AnyObject aliceServices = clientSession->service("AliceServices");
  alice::ImageStorePtr imageStore = aliceServices.call<alice::ImageStorePtr>("imageStore");
  assert(imageStore);
  storeImage(imageStore, imageName, imagePath);
  processImage(imageStore, imageName, "./tempfile");
  processImageWithProgress(imageStore, imageName, "./tempfile");
}
}

void setupAliceServer(qi::SessionPtr serverSession)
{
  qi::DynamicObjectBuilder objectBuilder;
  objectBuilder.advertiseMethod("imageStore", &alice::getImageStore);
  qi::AnyObject aliceServices = objectBuilder.object();
  serverSession->registerService("AliceServices", aliceServices);
}

int main(int argc, char** argv)
{
  ::TestMode::forceTestMode(TestMode::Mode_SD);
  qi::Application app(argc, argv);

  if (argc != 3)
    return EXIT_FAILURE;

  const auto imageFilePath = qi::Path::fromNative(argv[1]);
  const auto imageID = qi::Path::fromNative(argv[2]).str();
  TestSessionPair sessionPair;
  setupAliceServer(sessionPair.server());
  bob::doSomeWork(sessionPair.client(), imageFilePath, imageID);

  return EXIT_SUCCESS;
}
